### ISSUE TITLE
lxc-bridge-proxy fails when apt-get fetches multiple files

### ISSUE THREAD

**Issue by candlerb** [lxc/lxd#2166]
# Required information

 * Distribution: ubuntu
 * Distribution version: 16.04
 * The output of "lxc info":

~~~
apicompat: 0
auth: trusted
environment:
  addresses: []
  architectures:
  - x86_64
  - i686
  certificate: <snip>
  driver: lxc
  driverversion: 2.0.1
  kernel: Linux
  kernelarchitecture: x86_64
  kernelversion: 4.4.0-28-generic
  server: lxd
  serverpid: 2510
  serverversion: 2.0.2
  storage: zfs
  storageversion: "5"
config:
  images.auto_update_cached: "false"
  images.auto_update_interval: "0"
  storage.zfs_pool_name: lxd
public: false
~~~

(This is actually running inside a KVM VM, but I don't think that matters here)

# Issue description

Scenario: a new 16.04 container is created and it is relying on the default networking (eth0 to lxdbr0, `http_proxy=http://[fe80::1%eth0]:13128`)

Running an `apt-get install` which loads a single file works, but one which loads multiple files fails after the first one; it gets back 500 Internal Server Error.

# Steps to reproduce

The outer host is called "vtp" and the container is called "foo"

~~~
root@vtp:~# lxc launch images:ubuntu/xenial/amd64 foo
Creating foo
Starting foo
root@vtp:~# lxc exec foo /bin/bash
root@foo:~# apt-get update
Hit:1 http://archive.ubuntu.com/ubuntu xenial InRelease
Get:2 http://archive.ubuntu.com/ubuntu xenial-updates InRelease [94.5 kB]
Get:3 http://security.ubuntu.com/ubuntu xenial-security InRelease [94.5 kB]
Get:4 http://archive.ubuntu.com/ubuntu xenial-updates/main amd64 Packages [240 kB]
Get:5 http://archive.ubuntu.com/ubuntu xenial-updates/universe amd64 Packages [128 kB]
Fetched 557 kB in 0s (800 kB/s)
Reading package lists... Done
root@foo:~# apt-get install mtr-tiny nmap strace
Reading package lists... Done
Building dependency tree
Reading state information... Done
The following additional packages will be installed:
  libblas-common libblas3 libicu55 liblinear3 liblua5.2-0 libpcap0.8 libpython-stdlib
  libpython2.7-minimal libpython2.7-stdlib libxml2 libxslt1.1 lua-lpeg ndiff python python-bs4
  python-chardet python-html5lib python-lxml python-minimal python-pkg-resources python-six python2.7
  python2.7-minimal sgml-base xml-core
Suggested packages:
  liblinear-tools liblinear-dev python-doc python-tk python-genshi python-lxml-dbg python-lxml-doc
  python-setuptools python2.7-doc binutils binfmt-support sgml-base-doc debhelper
The following NEW packages will be installed:
  libblas-common libblas3 libicu55 liblinear3 liblua5.2-0 libpcap0.8 libpython-stdlib
  libpython2.7-minimal libpython2.7-stdlib libxml2 libxslt1.1 lua-lpeg mtr-tiny ndiff nmap python
  python-bs4 python-chardet python-html5lib python-lxml python-minimal python-pkg-resources
  python-six python2.7 python2.7-minimal sgml-base strace xml-core
0 upgraded, 28 newly installed, 0 to remove and 0 not upgraded.
Need to get 18.9 MB of archives.
After this operation, 79.1 MB of additional disk space will be used.
Do you want to continue? [Y/n] y
Get:1 http://archive.ubuntu.com/ubuntu xenial/main amd64 sgml-base all 1.26+nmu4ubuntu1 [12.5 kB]
Err:2 http://archive.ubuntu.com/ubuntu xenial/main amd64 libpython2.7-minimal amd64 2.7.11-7ubuntu1
  500  Internal Server Error
Err:3 http://archive.ubuntu.com/ubuntu xenial/main amd64 python2.7-minimal amd64 2.7.11-7ubuntu1
  500  Internal Server Error
Err:4 http://archive.ubuntu.com/ubuntu xenial/main amd64 python-minimal amd64 2.7.11-1
  500  Internal Server Error
Err:5 http://archive.ubuntu.com/ubuntu xenial/main amd64 libpython2.7-stdlib amd64 2.7.11-7ubuntu1
  500  Internal Server Error
Err:6 http://archive.ubuntu.com/ubuntu xenial/main amd64 python2.7 amd64 2.7.11-7ubuntu1
  500  Internal Server Error
Err:7 http://archive.ubuntu.com/ubuntu xenial/main amd64 libpython-stdlib amd64 2.7.11-1
  500  Internal Server Error
Err:8 http://archive.ubuntu.com/ubuntu xenial/main amd64 python amd64 2.7.11-1
  500  Internal Server Error
Err:9 http://archive.ubuntu.com/ubuntu xenial/main amd64 libicu55 amd64 55.1-7
  500  Internal Server Error
Err:10 http://archive.ubuntu.com/ubuntu xenial/main amd64 libpcap0.8 amd64 1.7.4-2
  500  Internal Server Error
Err:11 http://security.ubuntu.com/ubuntu xenial-security/main amd64 libxml2 amd64 2.9.3+dfsg1-1ubuntu0.1
  500  Internal Server Error
Err:12 http://archive.ubuntu.com/ubuntu xenial-updates/main amd64 mtr-tiny amd64 0.86-1ubuntu0.1
  500  Internal Server Error
Err:13 http://archive.ubuntu.com/ubuntu xenial/main amd64 strace amd64 4.11-1ubuntu3
  500  Internal Server Error
Err:14 http://archive.ubuntu.com/ubuntu xenial/main amd64 xml-core all 0.13+nmu2
  500  Internal Server Error
Err:15 http://archive.ubuntu.com/ubuntu xenial/main amd64 libblas-common amd64 3.6.0-2ubuntu2
  500  Internal Server Error
Err:16 http://archive.ubuntu.com/ubuntu xenial/main amd64 libblas3 amd64 3.6.0-2ubuntu2
  500  Internal Server Error
Err:17 http://archive.ubuntu.com/ubuntu xenial/main amd64 liblinear3 amd64 2.1.0+dfsg-1
  500  Internal Server Error
Err:18 http://archive.ubuntu.com/ubuntu xenial/main amd64 liblua5.2-0 amd64 5.2.4-1ubuntu1
  500  Internal Server Error
Err:19 http://archive.ubuntu.com/ubuntu xenial/main amd64 libxslt1.1 amd64 1.1.28-2.1
  500  Internal Server Error
Err:20 http://archive.ubuntu.com/ubuntu xenial/main amd64 lua-lpeg amd64 0.12.2-1
  500  Internal Server Error
Get:11 http://security.ubuntu.com/ubuntu xenial-security/main amd64 libxml2 amd64 2.9.3+dfsg1-1ubuntu0.1 [696 kB]
Err:21 http://archive.ubuntu.com/ubuntu xenial/main amd64 python-bs4 all 4.4.1-1
  500  Internal Server Error
Err:22 http://archive.ubuntu.com/ubuntu xenial/main amd64 python-pkg-resources all 20.7.0-1
  500  Internal Server Error
Err:23 http://archive.ubuntu.com/ubuntu xenial/main amd64 python-chardet all 2.3.0-2
  500  Internal Server Error
Err:24 http://archive.ubuntu.com/ubuntu xenial/main amd64 python-six all 1.10.0-3
  500  Internal Server Error
Err:25 http://archive.ubuntu.com/ubuntu xenial/main amd64 python-html5lib all 0.999-4
  500  Internal Server Error
Err:26 http://archive.ubuntu.com/ubuntu xenial/main amd64 python-lxml amd64 3.5.0-1build1
  500  Internal Server Error
Err:27 http://archive.ubuntu.com/ubuntu xenial/main amd64 ndiff all 7.01-2ubuntu2
  500  Internal Server Error
Get:28 http://archive.ubuntu.com/ubuntu xenial/main amd64 nmap amd64 7.01-2ubuntu2 [4638 kB]
Fetched 5346 kB in 2s (1819 kB/s)
E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/p/python2.7/libpython2.7-minimal_2.7.11-7ubuntu1_amd64.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/p/python2.7/python2.7-minimal_2.7.11-7ubuntu1_amd64.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/p/python-defaults/python-minimal_2.7.11-1_amd64.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/p/python2.7/libpython2.7-stdlib_2.7.11-7ubuntu1_amd64.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/p/python2.7/python2.7_2.7.11-7ubuntu1_amd64.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/p/python-defaults/libpython-stdlib_2.7.11-1_amd64.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/p/python-defaults/python_2.7.11-1_amd64.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/i/icu/libicu55_55.1-7_amd64.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/libp/libpcap/libpcap0.8_1.7.4-2_amd64.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/m/mtr/mtr-tiny_0.86-1ubuntu0.1_amd64.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/s/strace/strace_4.11-1ubuntu3_amd64.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/x/xml-core/xml-core_0.13+nmu2_all.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/l/lapack/libblas-common_3.6.0-2ubuntu2_amd64.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/l/lapack/libblas3_3.6.0-2ubuntu2_amd64.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/libl/liblinear/liblinear3_2.1.0+dfsg-1_amd64.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/l/lua5.2/liblua5.2-0_5.2.4-1ubuntu1_amd64.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/libx/libxslt/libxslt1.1_1.1.28-2.1_amd64.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/l/lua-lpeg/lua-lpeg_0.12.2-1_amd64.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/b/beautifulsoup4/python-bs4_4.4.1-1_all.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/p/python-setuptools/python-pkg-resources_20.7.0-1_all.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/c/chardet/python-chardet_2.3.0-2_all.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/s/six/python-six_1.10.0-3_all.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/h/html5lib/python-html5lib_0.999-4_all.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/l/lxml/python-lxml_3.5.0-1build1_amd64.deb  500  Internal Server Error

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/n/nmap/ndiff_7.01-2ubuntu2_all.deb  500  Internal Server Error

E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
~~~

Note that the first GET, and one of the later ones, are successful. But the rest give 500 errors.

If you do an operation which requires only fetching a single file, it works fine:

~~~
root@foo:~# apt-get install strace
Reading package lists... Done
Building dependency tree
Reading state information... Done
The following NEW packages will be installed:
  strace
0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 179 kB of archives.
After this operation, 913 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu xenial/main amd64 strace amd64 4.11-1ubuntu3 [179 kB]
Fetched 179 kB in 0s (1336 kB/s)
Selecting previously unselected package strace.
(Reading database ... 12917 files and directories currently installed.)
Preparing to unpack .../strace_4.11-1ubuntu3_amd64.deb ...
Unpacking strace (4.11-1ubuntu3) ...
Setting up strace (4.11-1ubuntu3) ...
root@foo:~#
~~~

So I suspect the problem might be to do with http pipelining from the client.

~~~
root@vtp:~# tcpdump -i lxdbr0 -nn -c100 -s0 -A tcp port 13128
...
13:16:16.551811 IP6 fe80::216:3eff:fe07:e76b.55948 > fe80::1.13128: Flags [P.], seq 180:1731, ack 28438, win 670, options [nop,nop,TS val 1960883 ecr 1960882], length 1551
`.w../.@..........>....k..................3H. &Y........)......
........GET http://archive.ubuntu.com/ubuntu/pool/main/p/python2.7/libpython2.7-stdlib_2.7.11-7ubuntu1_amd64.deb HTTP/1.1
Host: archive.ubuntu.com
User-Agent: Debian APT-HTTP/1.3 (1.2.12)

GET http://archive.ubuntu.com/ubuntu/pool/main/p/python2.7/python2.7_2.7.11-7ubuntu1_amd64.deb HTTP/1.1
Host: archive.ubuntu.com
User-Agent: Debian APT-HTTP/1.3 (1.2.12)

GET http://archive.ubuntu.com/ubuntu/pool/main/p/python-defaults/libpython-stdlib_2.7.11-1_amd64.deb HTTP/1.1
Host: archive.ubuntu.com
User-Agent: Debian APT-HTTP/1.3 (1.2.12)

GET http://archive.ubuntu.com/ubuntu/pool/main/p/python-defaults/python_2.7.11-1_amd64.deb HTTP/1.1
Host: archive.ubuntu.com
User-Agent: Debian APT-HTTP/1.3 (1.2.12)

GET http://archive.ubuntu.com/ubuntu/pool/main/i/icu/libicu55_55.1-7_amd64.deb HTTP/1.1
Host: archive.ubuntu.com
User-Agent: Debian APT-HTTP/1.3 (1.2.12)

GET http://archive.ubuntu.com/ubuntu/pool/main/libp/libpcap/libpcap0.8_1.7.4-2_amd64.deb HTTP/1.1
Host: archive.ubuntu.com
User-Agent: Debian APT-HTTP/1.3 (1.2.12)

GET http://archive.ubuntu.com/ubuntu/pool/main/m/mtr/mtr-tiny_0.86-1ubuntu0.1_amd64.deb HTTP/1.1
Host: archive.ubuntu.com
User-Agent: Debian APT-HTTP/1.3 (1.2.12)

GET http://archive.ubuntu.com/ubuntu/pool/main/x/xml-core/xml-core_0.13%2bnmu2_all.deb HTTP/1.1
Host: archive.ubuntu.com
User-Agent: Debian APT-HTTP/1.3 (1.2.12)

GET http://archive.ubuntu.com/ubuntu/pool/main/l/lapack/libblas-common_3.6.0-2ubuntu2_amd64.deb HTTP/1.1
Host: archive.ubuntu.com
User-Agent: Debian APT-HTTP/1.3 (1.2.12)


13:16:16.551838 IP6 fe80::1.13128 > fe80::216:3eff:fe07:e76b.55948: Flags [.], ack 1731, win 256, options [nop,nop,TS val 1960883 ecr 1960883], length 0
`.... .@..........................>....k3H....... ,h....#......
........
~~~

I would like to turn on some debugging from lxd-bridge-proxy but I can't see an obvious way to do so.

**Comment by stgraber** [comment]
Hmm, yeah, sounds like it could be a http pipelining issue.
Our code is rather trivial though so I'm not sure what we can do to fix this: https://github.com/lxc/lxd/blob/master/lxd-bridge/lxd-bridge-proxy/main.go

**Comment by candlerb** [comment]
I found a workaround for use inside the container: <http://bnpcs.blogspot.co.uk/2010/11/disable-http-pipelining-in-ubuntu.html>

This fixes it, so it's definitely a problem with pipelining.

However it would be good if we could get the relevant go http module fixed (i.e. ReverseProxy or ListenAndServe, whichever is at fault)

**Comment by stgraber** [comment]
So using an official ubuntu image (ubuntu:16.04), we don't have that problem because pipelining is disabled.

The Go side of this problem has been fixed upstream and will be in Go 1.7, then hopefully pushed as a bugfix to the other stable Go releases.

I did poke a bit at working around this in the lxd proxy code itself, but short of copying the whole upstream httputil library or making the whole proxy single-threaded 

So I'm going to consider this resolved by upstream https://github.com/golang/go/commit/4b0bc7c3a14ac446bc13d22098de8db382205401

**Comment by candlerb** [comment]
Thank you. That resolution is fine.

I have now switched to using the official images, which contain `/etc/apt/apt.conf.d/90cloud-init-pipelining` - I did not realise before that the images on linuxcontainers.org were different.
